com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs

```diff
@@ -8,20 +10,79 @@ namespace UnityEditor.ShaderFoundry
         readonly string m_AdditionalShaderID;
         readonly List<TemplateInstance> m_TemplateInstances;
         readonly string m_FallbackShader;
+        readonly List<ShaderCustomEditor> m_CustomEditors;
+        readonly List<ShaderDependency> m_Dependencies;
 
         public string Name => m_Name;
         public bool IsPrimaryShader => string.IsNullOrEmpty(m_AdditionalShaderID);
-        public IEnumerable<TemplateInstance> TemplateInstances => m_TemplateInstances;
+        public IEnumerable<TemplateInstance> TemplateInstances => m_TemplateInstances?.AsReadOnly() ?? Enumerable.Empty<TemplateInstance>();
         public string FallbackShader => m_FallbackShader;
+        public IEnumerable<ShaderCustomEditor> CustomEditors => m_CustomEditors?.AsReadOnly() ?? Enumerable.Empty<ShaderCustomEditor>();
+        public IEnumerable<ShaderDependency> Dependencies => m_Dependencies?.AsReadOnly() ?? Enumerable.Empty<ShaderDependency>();
+
         public bool IsValid => !string.IsNullOrEmpty(m_Name);
-        public static ShaderInstance Invalid => new ShaderInstance(null, null, null, null);
+        public static ShaderInstance Invalid => new ShaderInstance(null, null, null);
 
-        internal ShaderInstance(string name, string additionalShaderID, List<TemplateInstance> templateInstances, string fallbackShader)
+        internal ShaderInstance(string name, string additionalShaderID, List<TemplateInstance> templateInstances)
         {
             m_Name = name;
             m_AdditionalShaderID = additionalShaderID;
             m_TemplateInstances = templateInstances;
-            m_FallbackShader = fallbackShader;
+
+            // these are copied out of the Templates in the template instances
+            m_FallbackShader = null;
+            var customEditors = new List<ShaderCustomEditor>();
+            var dependencies = new List<ShaderDependency>();
+            foreach (var templateInstance in templateInstances)
+            {
+                if (!string.IsNullOrEmpty(templateInstance.Template.ShaderFallback))
+                {
+                    if (m_FallbackShader == null)
+                        m_FallbackShader = templateInstance.Template.ShaderFallback;
+                    else
+                    {
+                        if (m_FallbackShader != templateInstance.Template.ShaderFallback)
+                        {
+                            // TODO @ SHADERS ERROR - conflicting shader fallbacks defined...
```

Comment:
You may consider using `ErrorHandling.ReportError` or adding to this so that we have a unified place to update later instead of a bunch of separate ones.